src/utils/binding_utils.ts

```diff
@@ -1,6 +1,54 @@
-import { interfaces } from "../inversify";
-
+import { getServiceIdentifierAsString, interfaces } from "../inversify";
+import * as ERROR_MSGS from "../constants/error_msgs";
 export const multiBindToService = (container: interfaces.Container) =>
     (service: interfaces.ServiceIdentifier<any>) =>
         (...types: interfaces.ServiceIdentifier<any>[]) =>
             types.forEach((t) => container.bind(t).toService(service));
+
+export const _ensureFullyBound = (binding:interfaces.Binding<unknown>):void => {
+    let boundValue:unknown = null;
+    switch(binding.type){
+        case "ConstantValue":
+        case "Function":
+            boundValue = binding.cache;
+            break;
+        case "Constructor":
+        case "Instance":
+            boundValue = binding.implementationType;
+            break;
+        case "DynamicValue":
+            boundValue = binding.dynamicValue;
+            break;
+        case "Provider":
+            boundValue = binding.provider;
+            break;
+        case "Factory":
+            boundValue = binding.factory;
+            break;
+    }
+    if (boundValue === null) {
+            // The user probably created a binding but didn't finish it
+        // e.g. container.bind<T>("Something"); missing BindingToSyntax
+        const serviceIdentifierAsString = getServiceIdentifierAsString(binding.serviceIdentifier);
+        throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifierAsString}`);
+    }
+}
+export type FactoryType = "toDynamicValue" | "toFactory" | "toAutoFactory" | "toProvider";
+export type FactoryTypeFunction = (context: interfaces.Context) => any;
+
+export interface FactoryDetails {
+    factoryType:FactoryType,
+    factory:FactoryTypeFunction | null
+};
+export const _getFactoryDetails = (binding:interfaces.Binding<unknown>): FactoryDetails | null => {
+    switch(binding.type){
+        case "Factory":
+            return { factory: binding.factory,factoryType:"toFactory"};
+        case "Provider":
+            return { factory: binding.provider,factoryType:"toProvider"};
+        case "DynamicValue":
+            return { factory: binding.dynamicValue,factoryType:"toDynamicValue"};
+    }
+    return null;
+}
```

Comment:
Nice, but this function returns null as unexpected value. I think its cleaner to throw an exception:

```ts

export const _getFactoryDetails = (binding: interfaces.Binding<unknown>): FactoryDetails => {
    switch(binding.type){
        case BindingTypeEnum.Factory:
            return { factory: binding.factory, factoryType: FactoryType.Factory };
        case BindingTypeEnum.Provider:
            return { factory: binding.provider, factoryType: FactoryType.Provider };
        case BindingTypeEnum.DynamicValue:
            return { factory: binding.dynamicValue, factoryType: FactoryType.DynamicValue };
        default:
            throw new Error(`Unexpected factory type ${binding.type}`);
    }
}

```

This way you don't need to cast `_getFactoryDetails(binding) as FactoryDetails` later

Reply:
Agreed.